Master Stiffness Matrix Process

SADPR opened the pull request:
Master Stiffness Matrix Process:
Performs 6 differents simulations to obtain the stiffness matrix for a master node with respect to a given slave surface.

Still need to improve the given data such as Value (needs to be automatic from a characteristic L of the model Part), Master_coor, slave surface name. 

Sebastian Ares de Parga Regalado

Commit:
Master Stiffness Matrix Process

loumalouomega commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+        #Iterating through the entire list of constraints 
+        for j in range(self.parameters["processes"]["constraints_process_list"].size()):
+            constraints_process_list = self.parameters["processes"]["constraints_process_list"][j] 
+            if constraints_process_list["Parameters"]["model_part_name"].GetString()=="Structure.DISPLACEMENT_"+self.Slave_name: # Enforcing to change the displacement vector only in slave surface
```
Please don do this. Get the variable, and compare variables

loumalouomega commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+        for j in range(self.parameters["processes"]["constraints_process_list"].size()):
+            constraints_process_list = self.parameters["processes"]["constraints_process_list"][j] 
+            if constraints_process_list["Parameters"]["model_part_name"].GetString()=="Structure.DISPLACEMENT_"+self.Slave_name: # Enforcing to change the displacement vector only in slave surface
+                if DoF==0:
```
Please don't do this, use the variables for that

loumalouomega commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+                    constraints_process_list["Parameters"]["value"].SetVector([0.0, self.Value, 0.0]) #Setting the new vector of displacements.
+                elif DoF==2:
+                    constraints_process_list["Parameters"]["value"].SetVector([0.0, 0.0, self.Value]) #Setting the new vector of displacements.
+                elif DoF==3:
```
I don't like this level of hardcoding

RiccardoRossi replied:
i agree, you can simply reset the vector to zero and assign only the i-th component

loumalouomega commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+
+        dim = self.parameters["solver_settings"]["domain_size"].GetInt() #Dimension 
+
+        r = np.zeros((len(slave_undeformed_coordinates),dim)) # Initialize Position vector r 
```
Please do not add dependences with numpy if is not strictly necessary

loumalouomega commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+            self.MasterStiffnessVector()
+            self.Master_Stiffness[:,DoF] = self.Stiffness #Assembling of Master Stiffness Matrix
+        
+        print("\n \n The Stiffness Matrix related to the master node is: \n", self.Master_Stiffness)
```
Use the logger

loumalouomega reviewed (changes requested):
Comments

philbucher reviewed (changes requested):
Overall this looks good to me, but I have some comments that I will write later

In the meantime can you please add a small test?

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+        """
+        KratosMultiphysics.Process.__init__(self)
+        self.parameters = parameters
+        self.Value = Value
```
why don't you get those values from the parameter?

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+        self.parameters = parameters
+        self.Value = Value
+        self.Master_coor = np.array(Master_coor)
+        self.Stiffness = np.empty((6,1)) #Initialize the Stiffness column vector
```
use a Kratos Vector

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,126 @@
+                    constraints_process_list["Parameters"]["value"].SetVector([0.0, 0.0, self.Value]) #Setting the new vector of displacements.
+                elif DoF==3:
+                    constraints_process_list["Parameters"]["value"][0].SetDouble(0.0) #Since for the rotation we give double values and strings. We have to Set them separately
+                    constraints_process_list["Parameters"]["value"][1].SetString("((y-"+str(self.Master_coor[0,1])+")*cos("+str(self.Value)+")-(z-"+str(self.Master_coor[0,2])+")*sin("+str(self.Value)+"))+("+str(self.Master_coor[0,1])+"-y)")
```
my impression is that here it will be easier to directly compute the rotation matrix as a matrix, and to apply it as a condition to the nodes... something on the line of

~~~py
R = ComputeRotationMatrix(theta)
v = Vector(3)
for node in mp.Nodes:
    v[0] = node.X0 - master_coord[0]
    v[1] = node.Y0 - master_coord[0]
    v[2] = node.Z0 - master_coord[0]
    displacement = R*v
    node.SetSolutionStepValue(DISPLACEMENT)
    node.Fix(DISPLACEMENT_X)
    node.Fix(DISPLACEMENT_Y)
    node.Fix(DISPLACEMENT_Z)
~~~

Commit:
Now it is able to perform stiffness matrix for two master nodes and has no dependency of numpy.

Commit:
Takes all process' parameters as Kratos Parameters and a model.

Still need to change the hard-coded part for the forced Displacements, by assigning only the ith component and use the Rotation matrix... working on this.

Commit:
List of changes:
- Has KratosMultiphysics.Process structure.

- Validates and assign default parameters.

- The displacements are assigned to the ith component and the rotations are assign by calculating the rotation matrix.

- Doesn't perform 12 differents simulations anymore but 12 different steps.

- Note: there are substantial changes in the process.

Commit:
Small test for Master Stiffness Process.

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,203 @@
+
+        #Cross product (Obtaining moments)
+        for i in range(slave_num_nodes):
+            moments[i,0] = r[i,1]*slave_reaction[i,2]-r[i,2]*slave_reaction[i,1] #Moments x
```
I would have created an auxiliary function for this

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/master_stiffness_process.py:
```diff
@@ -0,0 +1,203 @@
+        KratosMultiphysics.Process.__init__(self)
+        
+        default_parameters = KratosMultiphysics.Parameters("""{
+                "model_part_list": ["Surface_name_1","Surface_name_2"],
```
absolutely cosmetic comment, but i would do it as 

~~~json
[
    { 
      "model_part_name"   : "Surface_name_1",
      "master_coordinates" : [0.0,0.0,0.0]
    },
   { 
      "model_part_name"   : "Surface_name_2",
      "master_coordinates" : [1.0,0.0,0.0]
    }
]
~~~

simply because this way you can read at one the information block you need.

other than that, well done!!

RiccardoRossi replied:
well, i think you should also have the "relative_perturbation_eps" to be user controllable

~~~json
[
    { 
      "model_part_name"   : "Surface_name_1",
      "master_coordinates" : [0.0,0.0,0.0],
      "eps_perturbation" : 1e-5
    },
   { 
      "model_part_name"   : "Surface_name_2",
      "master_coordinates" : [1.0,0.0,0.0],
        "eps_perturbation" : 1e-5
    }
]
~~~

SADPR replied:
I would like to leave the parameters as a list due to the logic of the algorithm, but please let me know if it is necessary and I change it.

RiccardoRossi reviewed (commented):
almost ready from my point of view

Commit:
Cross product as a function.
Added eps_perturbation parameter.
Update of small test parameters.

Note: I would like to leave the parameters as a list due to the logic of the algorithm, but please let me know if it is necessary and I change it.

Commit:
Created files:
- master_stiffness_process.cpp
- master_stiffness_process.h
/// Migrate the process from python to c++.
/// Previously the process only worked for two slave surfaces, now it works for "n" slave surfaces.

Created folder:
-master_stiffness_process_test
///Only created a proposed cheap test folder, didn't touch the test files.

Modified files:
- master_stiffness_process.py
/// Now only calls the c++ process.

- add_custom_processes_to_python.cpp
/// Added the link between python and c++.

Commit:
Minor changes

Small correction in master stiffness matrix.
Prints the Stiffness Matrices in a "Master_Stiffness_Matrices.json" file.

Commit:
One master stiffness matrix instead of n (where n stands for number of slave surfaces).

Commit:
Added quaternion rotation matrix and master-slave constraints.

Commit:
Modified output paramters on json file

Commit:
Stress and Reaction ROM

Commit:
Stress and Reaction All files with tests

Commit:
Funciona hasta tomar las snapshots de residuales de los elementos restringidos en HROM, queda pendiente hacer la reconstrucción dado el arreglo HR_residuals

Commit:
Reconstructs the residual when restricted elements appear on the HROM model, need to pass a list to force at least n number of elements.

Commit:
Reconstruye los residuales de manera correcta

Commit:
Agrego empirical_cubature... que acepta una lista de elementos restringidos.
Agrego el strcutural_mechanics... donde paso la lista de los elementos restringidos que deseo incluir en mi modelo HROM

Commit:
Reconstruction of reactions for a constant tetrahedral mesh with 3 dof_node (12 dof_elem)

Commit:
First attempt to assemble stresses

Commit:
On folder SEBASTIAN_ROM_PK2, the reaction is reconstructed on the nodes and the value is directly set to the node solution.
The reconstruction of the stress is now for the pk2 field and the vtk for the stresses is created with the help of postprocessing.py.

Commit:
Erased unnecessary fields and commented some parts of the code. First step into cleaning the final code. Sparse matrix initialized.

Commit:
PK2_STRESS_VECTOR changed to nodal data value variables, no need to used postprocessing.py anymore

Commit:
Reaction Reconstruction process
Stress Reconstruction process (PK2_STRESS_VECTOR)
Added a folder with the mdpa, .py, .json to run a test.

Commit:
Added rom_basis_process.
Corrected the test folder.

Commit:
Only one RomParameters for the reconstruction.
Assemble stresses in c++.
Assemble reactions improved.
Commented section, work in progress for the stress reconstruction. Need to build the sparse matrix in c++.
Updated the test folder "Sebastian_HROM_pk2_reaction_with_process

Commit:
New builder and solver

Commit:
My configure file

Commit:
Need to check on hrom_stiffness_matrix, update boundary condition, and master stiffness matrix (static condensation)

Commit:
Need to check on hrom_stiffness_matrix, update boundary condition, and master stiffness matrix (static condensation)

Commit:
Small change
